keps/sig-node/20190626-node-shutdown-state.md

```diff
@@ -0,0 +1,253 @@
+4. register node as being processed in order to renew the Lease periodically
+5. delete the pod with a 0-grace-period
+7. once 5. returns untaint the node and stop renewing the Lease
+
+
+To avoid being too reactive in 2. we can backoff and not try to acquire the lock immediately to avoid evicting too fast
+
+#### Kubelet
+
+we need also to apply changes to the kubelet, at startup the kubelet will:
+
+1. acquire the Lease:
+  a. fails: do not start any pods, unless the pod tolerates the node shutdown taint
+  b. success: start normally
+
+### Risks and Mitigations
+
+This KEP introduces changes, and if not tested correctly could result in data corruption for users.
+To mitigate this we plan to have a high test coverage and to introduce this enhancement as an alpha
+
+## Design Details
+
+### Test Plan
+
+This feature will be tested with the following approaches:
+
+- unit tests
+- integration
+- e2e
+
+we also plan to test this with different version Skews.
+
+### Graduation Criteria
+
+This KEP will be treated as a new feature, and will be introduced with a new feature gate, NodeShutdownFailover.
+
+This enhancement will go through the following maturity levels: alpha, beta and stable.
+
+Graduation criteria between these levels to be determined.
+
+### Upgrade / Downgrade Strategy
+
+If applicable, how will the component be upgraded and downgraded? Make sure this is in the test plan.
+
+Consider the following in developing an upgrade/downgrade strategy for this enhancement:
+- to keep the old behavior: disable the feature gate
+- make sure that the new behavior is not running against lower-version Kubelets
+
+### Version Skew Strategy
+
+This feature requires version parity between the control plane and the kubelet, this behavior shouldn't be enabled in case of older Kubelets
+or Kubelets disabling the feature gate. To do this, we will make the kubelet add annotation `lease.kubernetes.io/node-failover=true` 
+to let the controller know that the feature is enabled on the kubelet-side.
+
+
+## Implementation History
+
+- 2019-06-26: Initial KEP published.
+
+## Drawbacks
+
+This design adds a new traffic into the apiserver, we should make sure that this feature do not break our SLOs
+
+## Alternatives
+
+In order to achieve failover properties, the control plane needs to coordinate with Kubelet.
+
+### rely on the taint as a lock mechanism (taint and untaint by the same actor)
+
+1. gc_controller taints the node objects when it observe a node is shutdown
+2. gc_controller starts removing forcefully the pods
+3. gc_controller untaint the node when the eviction is done
+
+the kubelet will:
+
+1. start and fetch the node object from the apiserver
+  1.a. fails: doesn't start containers unless it tolerates the taint
+  1.b. success
+2. check for the taint
+  2.a. exists: same as 1.a
+  2.b. doesn't exist: start normally
+
+This solution has many scenarios to consider and can be error-prone, this is why we added the lease as Locking
```

Comment:
Does "can be error-prone" reference the race condition of setting the taint and the kubelet running pods?  Or is there something else being considered?

Reply:
"Error prone" refers here to the fact that with this alternative we would be re-implementing lock semantics around the taint